velox/expression/CastHooks.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/StringWriter.h"
+
+namespace facebook::velox::exec {
+
+/// This class provides cast hooks to allow different behaviors of CastExpr and
+/// SparkCastExpr. The main purpose is crate a Spark-compatible cast
+/// implementation by taking full usage of existing cast expression. The default
+/// logics are following Presto semantics.
+class CastHooks {
+ public:
+  CastHooks(const core::QueryConfig& config)
+      : legacyCast_(config.isLegacyCast()) {}
```

Comment:
I wonder if it would be cleaner to have pure abstract CastHooks class and 2 separate implementations: PrestoCastHooks and SparkCastHooks. What do you think?

For example, the constructors logic for setting legacyCast_  from config is Presto-specific and it is a bit strange to run it as part of SparkCastHooks.

Reply:
Thanks for the suggestion. Added PrestoCastHooks and converted CastHooks to be pure abstract.